### PR TITLE
Include google search in sidebar

### PR DESCRIPTION
**Pull request by mjibson**
Looks like:

search

Search result page:

search-res

### CODE REVIEW

**Comment by jess-edwards** [comment]
@mjibson Is the left side nav the most intuitive place for a search box? At the very least, I think the search box should be on top of the left side nav, but possibly in the top right of the docs window. 

@jseldess Thoughts?

**Comment by mjibson** [comment]
I don't care where it goes. The place where it went (but was commented out) was broken, so I put it in the sidebar because it was easy and looked fine (though maybe not great).

I don't think I agree about putting it on the top of the sidenav on the left. I don't recall seeing that in any place, and I think it will mostly get in the way of people who want to click around in the nav. I think it should go in the top nav, but that nav doesn't really have space for a search box up there.

**Comment by mjibson** [comment]
For comparison, it's in the top nav on these sites:

https://angularjs.org/
http://www.postgresql.org/docs/9.4/static/index.html
https://sqlite.org/

It's at the top of the sidebar on these sites:

https://dev.mysql.com/doc/refman/5.7/en/

No search bar:

https://facebook.github.io/react/docs/getting-started.html
https://www.rethinkdb.com/docs/

**Comment by jseldess** [comment]
This is great, Matt! 

I'm not sure of the best place, but I worry that the spot below the sidenav is not prominent enough (once the docs are bigger, large majority will start off by searching). I like having it in the topnav (maybe in a collapse-like initial state), but I've seen it at the top of the sidenav quite often. Aerospike does both: http://www.aerospike.com/docs/

@mjibson, can you try it at the top of the sidenav so we can take a look? 

And I wonder if we shouldn't limit the scope even further to "cockroachlabs.com/docs"? I guess it's not bad to have blog posts in the results, but I don't know if we'll want other parts of the company website showing up for these searches... 

Also, I see that old, cached versions of pages are showing up in the results. Is there any way to limit the search to current versions?

**Comment by bdarnell** [comment]
We might be able to use google's custom search feature (http://cse.google.com) to control what shows up in the result, but I'm not sure if that can help with the old cached versions. We may have to host the search ourselves to have exact control over what's included (sphinx has a javascript-based search that is purely client side, although aside from the ability to control exactly which pages are covered the results aren't as good as google). Or we might be able to do something (a sitemap?) to get google to drop the stale pages faster. 

+1 to putting the search at the top of the sidebar (or the top nav if we can make space for it). http://docs.python.org is the only place where I've seen search at the bottom of the sidebar and I think I use it less because of the location. (it's also using the aforementioned sphinx-based search)

**Comment by jseldess** [comment]
This js-based search looks interesting: https://github.com/christian-fei/Simple-Jekyll-Search
Can see a demo here: https://www.npmjs.com/package/simple-jekyll-search

And here's the one called out in [Jekyll's own docs](http://jekyll.tips/tutorials/search/): http://lunrjs.com/

I think we should aim for our own hosted version, long-term, to be able to control the experience more and keep the results Cockroach-branded. For beta, though, Google could be ok. Should get it in above the sidebar and ask people to test it out?

**Comment by mjibson** [comment]
@jseldess the simple-jekyll-search module is exactly what this PR is removing. It works by populating a search.json file which is fetched and indexed in memory by the browser. The search search javascript searches that and tries to show some good results. I think doing client-side search is wrong: it's slow, forces users to download the entire contents of the site into memory if they want to search, isn't that smart (i.e., doesn't do cool things like stemming and other tricks google knows about), and doesn't have a nice results page. I'm of the opinion that we do server-side or nothing.

@bdarnell What's the advantage of CSE over using `site:cockroachlabs.com` in the query? We can easily create a sitemap file if we're worried about old results. I think getting google to crawl us enough to have recent results is an easier problem to solve than implementing our own search for now.

Notably, the sqlite search uses their own full text search. I think we should use google search for now until we can provide our own.

**Comment by bdarnell** [comment]
A CSE lets you add a stylesheet to the result page so it doesn't look like you just linked out to google. It  has better control over what's included (for example, we could create a CSE that includes both /docs/ and /blog/ but not the company info pages, or other things that would be difficult or impossible to express with `site:` queries). You can also pay for a CSE to disable ads in the results.

**Comment by jseldess** [comment]
I'm fine using google for now, but I don't understand why you feel so
strongly about client-side. Seems very typical in docs and I don't think it
impacts the experience in a way that most readers notice (docs aren't an
application). The client-side search experience here, for example, is quite
fast and good: http://www.aerospike.com/docs/. And check out the search
results: It's much much nicer to provide those in the context of the docs
than taking them away to a Google results page.

**Comment by mjibson** [comment]
@jseldess Can you show me an example of a site that uses client side search? aerospike makes an API call into google to fetch those results, which is server side. I may not have explained client-side search in a clear way. I can discuss over chat if that would help.

**Comment by mjibson** [comment]
I feel so strongly about client side because I think it's a horrible experience for users. It's slow because it requires them downloading the ENTIRE site contents into their browser, which uses up lots of bandwidth (slow, realllly slow on mobile or anywhere not near the data center). It has high memory and CPU requirements because it has to index or search the entire contents in some way (again, really bad for phones or less powerful computers which don't have a lot of that). It doesn't allow for google-like smarts (stemming, word misspellings, N-grams of words near each other, other special sauce I have no idea about) unless those smarts are implemented in the javascript, which they aren't.

In summary: server side is faster, less resource intensive for users, and provides better results.

**Comment by jseldess** [comment]
OK, these are good points. I jumped to conclusions. It's likely that sites I'm thinking of are searching server-side. Thanks for clarifying. 

Here are a few more that I assumed were client-side (they all provide results in the frame of the docs):
- http://docs.cartodb.com/cartodb-platform/sql-api/
- http://docs.basho.com/riak/latest/
- https://docs.djangoproject.com/en/1.9/
- https://support.stripe.com/

I think the important open point, then, is eventually having more control over the look and feel of the results and what gets included.

**Comment by mjibson** [comment]
Moved to the top of the sidebar. Restricted results to `/docs`. Looks like this:

top

I tried out CSE and didn't enjoy it. It's either a weird dialog popup (blergh, no thanks), or it links you to a page that has the same results as the one I have here, it just looks worse. We could prototype a full CSE thing where we attempt to render the results in our own chrome. I'm not against doing this, but I think that can wait for another PR because it's a fair amount of work, and what we have here works and looks pretty great. If I hit the search box on a site and it takes me to a google results page, I'm happy because I know I'm getting high quality results.

**Comment by jseldess** [comment]
Nice, Matt. Can you add `padding-bottom: 22px;`? Looks a bit too close to **Get Started**. Otherwise, LGTM.

@jess-edwards: Once this is merged, we can put it on DivisionOf's backlog to think about tweaking the search form styling a bit.

**Comment by mjibson** [comment]
Done.
